### PR TITLE
TEL-5477:  Unblocking the rtmp connections upon failure

### PR DESCRIPTION
**Pull request by mishehu**
The only way for this to work properly without flooding the logs until the b2bua keels over was to make a couple very small changes to the core.  The problem is that by the time we receive the file name in mod_av, it's already been stripped such that any calls to switch_ivr_stop_record_session() will fail to find the media bug in the channel's private_hash .

### CODE REVIEW

File: src/mod/applications/mod_av/avformat.c [hunk 1]
```diff
@@ -2229,6 +2227,29 @@ GCC_DIAG_ON(deprecated-declarations)
...
+					
+					if(ret == -(ETIMEDOUT)) {
+						context->closed = 1;
+						status = SWITCH_STATUS_FALSE;
+						
+						if(!zstr(context->owner_uuid)) {
+							switch_core_session_t *session = switch_core_session_locate(context->owner_uuid);
+						
+							if(session) {
+								if(!zstr(handle->original_file_path) && (ret = switch_ivr_stop_record_session(session, handle->original_file_path)) != SWITCH_STATUS_SUCCESS) {
```
Comment (dev-ryanc) [inline src/mod/applications/mod_av/avformat.c:2239]: I don't recommend calling `switch_ivr_stop_record_session` here. Take note that this is based on Freeswitch File Implementation, meaning I can also used this in playback. I would suggest to only return an error here, update the record_callback in switch_ivr_async and terminate when write fails. We should have a flipper as well to only do this if `RECORD_FAIL_ON_WRITE` is set to true. When we return error in media bug it should close the media bug by marking it as prune

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

**Review by Elhodred — APPROVED** [review]

File: src/mod/applications/mod_av/avformat.c (hunk 1 above)
Comment (mishehu) [reply src/mod/applications/mod_av/avformat.c:2239]: There is no need for a feature flipper.  If rw_timeout is not passed, then avio_open() is used, and the whole code path is moot.

I am not understanding the actual concern regarding the use of `switch_ivr_stop_record_session()` here.  Is there ever a situation where:

1.  We have an rtmp stream up that is being used for BOTH reading (playback) and writing (recording) AND
2.  Where if the socket timed out (note that this only currently happens on -(ETIMEOUT), meaning -110, for some reason they used the negative of the standard error value) for writing, that it somehow by magic the socket would still be up for reading (playback) ?  (This scenario is impossible, btw.)

Presumably (I didn't have a chance to check it yet), but upon calling `switch_ivr_stop_record_session()` , an event would be fired that call control could react upon in real time to re-issue a uuid_record command on that call.  (I've not received feedback from @bluesbettle  if that is the desired behavior or if it better to just set a chanvar that would indicate incomplete recording so post-processing could be done on the fullcap).

File: src/switch_ivr_async.c [hunk 2]
```diff
@@ -1404,6 +1404,9 @@ static void *SWITCH_THREAD_FUNC recording_thread(switch_thread_t *thread, void *
 				switch_channel_hangup(channel, SWITCH_CAUSE_DESTINATION_OUT_OF_ORDER);
 				switch_core_session_reset(session, SWITCH_TRUE, SWITCH_TRUE);
 			}
+			
+			// We might need to consider a threshold counter that we need to exceed first before we prune it
+			switch_set_flag(bug, SMBF_PRUNE);
```
Comment (dev-ryanc) [inline src/switch_ivr_async.c:1409]: We need to add that channel variable to immediately stop media bug on write fail similar to the statement above which check for `hangup_on_error` e.g. `stop_on_write_fail` or something

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

File: src/switch_ivr_async.c (hunk 2 above)
Comment (mishehu) [reply src/switch_ivr_async.c]: Flipper based upon chanvar in the code now.  RECORD_STOP_WRITE_ON_ERROR.

**Review by baloeng — APPROVED** [review]

**Pull request merged**